ApplicationInsightsRequestLogging/ApplicationInsightsRequestLogging/Filters/SensitiveDataFilter.cs

```diff
@@ -0,0 +1,125 @@
+        private const string SensitiveValueMask = "***MASKED***";
+
+        public readonly HashSet<string> _sensitiveDataPropertyKeys;
+        private readonly IEnumerable<string> _regexesForSensitiveValues;
+
+
+        public SensitiveDataFilter(BodyLoggerOptions options) : this(options.PropertyNamesWithSensitiveData, options.SensitiveDataRegexes)
+        {
+
+        }
+
+        public SensitiveDataFilter(IEnumerable<string> sensitiveDataPropertyKeys, IEnumerable<string> regexesForSensitiveValues)
+        {
+            _sensitiveDataPropertyKeys = sensitiveDataPropertyKeys.Select(t => t.ToLowerInvariant()).ToHashSet();
+            _regexesForSensitiveValues = regexesForSensitiveValues;
+        }
+
+        public string RemoveSensitiveData(string textOrJson)
+        {
+            try
+            {
+                var json = JsonNode.Parse(textOrJson);
+                if (json == null) return string.Empty;
+
+                if (json is JsonValue jValue && TestIfContainsSensitiveData("", jValue.ToString(), _sensitiveDataPropertyKeys, _regexesForSensitiveValues))
+                {
+                    return SensitiveValueMask;
+                }
+                RemoveIds(json);
+                return json.ToJsonString();
+            }
+            catch (JsonException)
+            {
+                if (TestIfContainsSensitiveData("", textOrJson, _sensitiveDataPropertyKeys, _regexesForSensitiveValues))
+                {
+                    return SensitiveValueMask;
+                }
+                return textOrJson;
+            }
+        }
+
+        private void RemoveIds(JsonNode node)
+        {
+            if (node is JsonObject jObject)
+            {
+                RemoveIds(jObject);
+            }
+            else if (node is JsonArray jArray)
+            {
+                RemoveFromArray(jArray);
+            }
+        }
+
+        private void RemoveIds(JsonObject? jObject)
+        {
+            if (jObject == null) throw new ArgumentNullException(nameof(jObject));
+
+            foreach (var jProperty in jObject.ToList())
+            {
+                if (jProperty.Value is JsonArray array)
+                {
+                    RemoveFromArray(array);
+                }
+                else if (jProperty.Value is JsonObject obj)
+                {
+                    RemoveIds(obj);
+                }
+                else if (jProperty.Value is JsonValue val
+                    && TestIfContainsSensitiveData(jProperty.Key, val.ToString(), _sensitiveDataPropertyKeys, _regexesForSensitiveValues))
+                {
+                    jObject[jProperty.Key] = SensitiveValueMask;
+                }
+            }
+        }
+
+        private void RemoveFromArray(JsonArray jArray)
+        {
+            if (jArray == null) throw new ArgumentNullException(nameof(jArray));
+
+            foreach (var jNode in jArray.Where(v => v != null))
+            {
+                RemoveIds(jNode);
```

Comment:
Actually, I am not clear about the implications of a possibly null reference argument here as raised by the compiler.